src/runner.rs

```diff
@@ -1252,65 +1108,92 @@ pub fn check_function_args(signature: &Signature, args: &[RuntimeValue]) -> Resu
 	Ok(())
 }
 
+#[derive(Debug)]
 struct ValueStack {
```

Comment:
Why `Box<[RuntimeValue]>` + `usize` and not `Vec`? This seems basically equivalent to a `Vec`.

Reply:
That's actually for a reason. This way it's more explicit that we don't use `Vec`, and in particular, that we are trying to avoid `Vec::push` method.

Reply:
I feel like that's unnecessary and just leads to having to write many `0`s into the `Box<[RuntimeValue]>`. If we used `Vec` we could have the extra space be transparently lazilly allocated by jemalloc

Reply:
AFAIR, using `Vec` in this case actually slows down quite a lot 

> I feel like that's unnecessary

What exactly? Do you mean avoiding using `Vec::push`? 

> just leads to having to write many 0s into the Box<[RuntimeValue]>

But if we use `Vec` we still need to use `resize`, which is basically the same?

Reply:
Why do we need to use `resize` instead of `with_capacity`? Surely one of the validation steps of the bytecode is to make sure that it doesn't read uninitialised stack data.

Reply:
If we use `with_capacity` only and leave `len` unchanged, then we basically can't use `get`, right?

Hm, I think I lost it somewhere and don't follow...

Reply:
I'm thinking something like this:

```rust
#[derive(Debug)]
struct ValueStack {
	buf: Vec<RuntimeValue>,
}

impl ValueStack {
	fn with_limit(limit: usize) -> ValueStack {
		let buf = Vec::with_capacity(limit);

		ValueStack {
			buf: buf,
		}
	}

	#[inline]
	fn drop_keep(&mut self, drop_keep: isa::DropKeep) {
		if drop_keep.keep == isa::Keep::Single {
			let top = *self.top();
			*self.pick_mut(drop_keep.drop as usize + 1) = top;
		}

		let cur_stack_len = self.len();
		self.buf.truncate(cur_stack_len - drop_keep.drop as usize);
	}

	#[inline]
	fn pop_as<T>(&mut self) -> T
	where
		T: FromRuntimeValue,
	{
		let value = self.pop();
		value.try_into().expect("Due to validation stack top's type should match")
	}

	#[inline]
	fn pop_pair_as<T>(&mut self) -> (T, T)
	where
		T: FromRuntimeValue,
	{
		let right = self.pop_as();
		let left = self.pop_as();
		(left, right)
	}

	#[inline]
	fn pop_triple(&mut self) -> (RuntimeValue, RuntimeValue, RuntimeValue) {
		let right = self.pop();
		let mid = self.pop();
		let left = self.pop();
		(left, mid, right)
	}

	#[inline]
	fn top(&self) -> &RuntimeValue {
		self.pick(1)
	}

	fn pick(&self, depth: usize) -> &RuntimeValue {
		&self.buf[self.buf.len() - depth]
	}

	#[inline]
	fn pick_mut(&mut self, depth: usize) -> &mut RuntimeValue {
	    let old_len = self.buf.len();
		&mut self.buf[old_len - depth]
	}

	#[inline]
	fn pop(&mut self) -> RuntimeValue {
	    // TODO: Use `get_unchecked` since we always have at least one value
		self.buf.pop().unwrap()
	}

	#[inline]
	fn push(&mut self, value: RuntimeValue) -> Result<(), TrapKind> {
	    if self.buf.len() == self.buf.capacity() {
	        return Err(TrapKind::StackOverflow);
	    }
	    self.buf.push(value);
		Ok(())
	}

	#[inline]
	fn len(&self) -> usize {
		self.buf.len()
	}
}
```

Reply:
I just re-run benchmarks with this version and they show that this version is indeed slower (not as slower as I expected though).